### PR TITLE
(SERVER-409) Plumb directories into jruby

### PR DESCRIPTION
**Pull request by nwolfe**
This commit adds plumbing for codedir, rundir, logdir in the same way we
were plumbing confdir and vardir through to JRuby.

This commit also bumps up the ruby Puppet submodule.

----

:+1: Unit tests pass
:+1: Can run the Server from source

:-1: Server is broken in packages, but not sure it's due to these changes, still investigating
**UPDATE**: Broken due to bindir and ssldir issues that are expected and not related to this PR

### CODE REVIEW

File: dev-resources/puppetlabs/services/jruby/jruby_interpreter_test/puppet.conf [hunk 1]
```diff
@@ -1,2 +1,2 @@
 [main]
-vardir = ./target/master-var
\ No newline at end of file
+vardir = ./target/master-var-jruby-int-test
```
Comment (jeffmccune) [inline dev-resources/puppetlabs/services/jruby/jruby_interpreter_test/puppet.conf]: Whitespace error, no trailing newline.

Comment (nwolfe) [inline dev-resources/puppetlabs/services/jruby/jruby_interpreter_test/puppet.conf]: It's not an error, but sure....thought my editor actually puts these in so not sure why this doesn't have one :smile:

Comment (jeffmccune) [inline dev-resources/puppetlabs/services/jruby/jruby_interpreter_test/puppet.conf]: "deliberate whitespace defect"

File: src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj [hunk 2]
```diff
@@ -176,39 +203,39 @@
    id       :- schema/Int
    config   :- JRubyPuppetConfig
    profiler :- (schema/maybe PuppetProfiler)]
-  (let [{:keys [ruby-load-path gem-home master-conf-dir master-var-dir
+  (let [{:keys [ruby-load-path gem-home
                 http-client-ssl-protocols http-client-cipher-suites]} config]
     (when-not ruby-load-path
-      (throw (Exception.
+      (throw (IllegalStateException.
```
Comment (nwolfe) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj]: Anything is better than an Exception - ISE seemed reasonable, but happy to change to something else

**Comment by jeffmccune** [comment]
Read through everything else and am :+1: 

I'd like to see how this behaves in the larger AIO context, but there's not a great way to know without merging this, so LGTM.

In the meantime, I'll try merging these changes into the AIO branch just for kicks and to get a sense of what the future holds.

**Comment by camlow325** [comment]
Since this PR would add a few new master-*-dir settings to the `jruby-puppet` section, I think it would be good to include some documentation for those in .../documentation/configuration.markdown and .../documentation/dev_running_from_source.markdown#step-3-set-up-config-files

**Comment by camlow325** [comment]
:+1: All looks really good to me - nice cleanup / consolidation in several places.  Only had the one comment about some additional documentation for the new settings.

**Comment by nwolfe** [comment]
> Since this PR would add a few new master-*-dir settings to the jruby-puppet section, I think it would be good to include some documentation for those in .../documentation/configuration.markdown and .../documentation/dev_running_from_source.markdown#step-3-set-up-config-files

I thought my documentation changes were a little light :smile: Will do...

**Comment by nwolfe** [comment]
@camlow325 Updated documentation

File: documentation/configuration.markdown [hunk 3]
```diff
@@ -60,15 +60,26 @@ This file contains the settings for Puppet Server itself.
...
     * `master-var-dir`: Optionally, set the path to the Puppet variable
       directory. If not specified, uses the Puppet default
-      `/opt/puppetlabs/puppet/cache`.
-    * `max-active-instances`: Optionally, set the maximum number of JRuby instances to allow. Defaults to 'num-cpus+2'.
-    * `borrow-timeout`: Optionally, set the timeout when attempting to borrow an instance from the JRuby pool in milliseconds. Defaults to 1200000.
+      `/opt/puppetlabs/server/data/puppetserver`.
+    * `master-run-dir`: Optionally, set the path to the Puppet run directory.
+      If not specified, uses the Puppet default `/var/run/puppetlabs/puppetserver`.
+    * `master-log-dir`: Optionally, set the path to the Puppet log directory.
+      If not specified, uses the Puppet default `/var/log/puppetserver/puppetserver`.
```
Comment (camlow325) [inline documentation/configuration.markdown]: `/var/log/puppetlabs/puppetserver`?

Comment (nwolfe) [inline documentation/configuration.markdown]: Dangit - fixed!

**Comment by camlow325** [comment]
:+1:

**Pull request merged**